### PR TITLE
Postprocess for keras model

### PR DESCRIPTION
**Pull request by mingxingtan**

### CODE REVIEW

File: efficientdet/keras/efficientdet_keras.py [hunk 1]
```diff
@@ -393,7 +394,9 @@ def __init__(self,
 
       bn_per_level = {}
       for level in range(self.min_level, self.max_level + 1):
-        bn_per_level[level] = utils_keras.build_batch_norm(
+        # NOTE(tanmingxing): key must be str to support model save/load.
```
Comment (fsx950223) [inline efficientdet/keras/efficientdet_keras.py:397]: Could we unified use array or dict?

File: efficientdet/keras/efficientdet_keras.py [hunk 2]
```diff
@@ -889,3 +891,51 @@ def call(self, features):
     box_outputs = self.box_net(feats)
 
     return class_outputs, box_outputs
+
+
+class EfficientDetModel(EfficientDetNet):
```
Comment (fsx950223) [inline efficientdet/keras/efficientdet_keras.py:896]: Awesome

File: efficientdet/keras/postprocess.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+  for i in range(batch_size):
+    nms_boxes, nms_scores, nms_classes = nms(params, boxes[i], scores[i],
+                                             classes[i])
+    nms_boxes = clip_boxes(nms_boxes, params['image_size'])
+
+    if batch_size > 1:
+      # pad to fixed length if batch size > 1.
+      max_output_size = params['nms_configs'].get('max_output_size',
+                                                  MAX_BOXES_PER_IMAGE)
+      nms_boxes = pad_zeros(nms_boxes, max_output_size)
```
Comment (fsx950223) [inline efficientdet/keras/postprocess.py:205]: Don't need pad_zeros.
We could use ```tf.raw_ops.NonMaxSuppressionV5(
         boxes = boxes,
         scores = scores,
         max_output_size = max_boxes_to_draw,
         iou_threshold=iou_threshold,
         score_threshold=min_score_thresh,
         soft_nms_sigma=soft_nms_sigma,
         pad_to_max_output_size=True)``` instead.

File: efficientdet/keras/postprocess.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    img_scales: scaling factor or the final image and bounding boxes.
+
+  Returns:
+    A tuple of batch level (boxes, scores, classess) after nms.
+  """
+  boxes, scores, classes = pre_nms(params, cls_outputs, box_outputs)
+
+  # A list of batched boxes, scores, and classes.
+  nms_boxes_bs, nms_scores_bs, nms_classes_bs = [], [], []
+  batch_size = boxes.get_shape()[0]
```
Comment (fsx950223) [inline efficientdet/keras/postprocess.py:195]: .shape

File: efficientdet/keras/postprocess.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+  boxes, scores, classes = pre_nms(params, cls_outputs, box_outputs)
+
+  nms_boxes_bs, nms_scores_bs, nms_classes_bs = [], [], []
+  batch_size = boxes.get_shape()[0]
+  for i in range(batch_size):
+    boxes_i, scores_i, classes_i = boxes[i], scores[i], classes[i]
+    nms_boxes_cls, nms_scores_cls, nms_classes_cls = [], [], []
+    for cid in range(params['num_classes']):
+      indices = tf.squeeze(tf.where(classes_i == cid))
+      if indices.get_shape().as_list() == 0:
```
Comment (fsx950223) [inline efficientdet/keras/postprocess.py:248]: .shape

File: efficientdet/keras/efficientdet_keras.py (hunk 1 above)
Comment (mingxingtan) [reply efficientdet/keras/efficientdet_keras.py]: I played with dict/array, and decide to use arrays for its simplicity.

File: efficientdet/keras/postprocess.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+MAX_BOXES_PER_IMAGE = 100
+
+
+def clip_boxes(boxes: T, image_size: int) -> T:
+  """Clip boxes to fit the image size."""
+  image_size = utils.parse_image_size(image_size) * 2
+  return tf.clip_by_value(boxes, [0], image_size)
+
+
+def pad_zeros(inputs: T, max_output_size: int, indices=None) -> T:
```
Comment (fsx950223) [inline efficientdet/keras/postprocess.py:34]: Could be removed.

File: efficientdet/keras/postprocess.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+    nms_boxes_cls = tf.concat(nms_boxes_cls, 0)
+    nms_scores_cls = tf.concat(nms_scores_cls, 0)
+    nms_classes_cls = tf.concat(nms_classes_cls, 0)
+    nms_valid_len_cls = tf.concat(nms_valid_len_cls, 0)
+    # get top detections and pad to fix size.
+    max_output_size = params['nms_configs'].get('max_output_size', 100)
+    _, indices = tf.math.top_k(nms_scores_cls, k=max_output_size, sorted=True)
+
+    nms_boxes_bs.append(pad_zeros(nms_boxes_cls, max_output_size, indices))
```
Comment (fsx950223) [inline efficientdet/keras/postprocess.py:267]: Could be removed.

**Comment by fsx950223** [comment]
I got bad result.

**Comment by mingxingtan** [comment]
> I got bad result.

Bug is fixed. You should be able to get the same outputs as non-keras models.

**Comment by fsx950223** [comment]
LGTM

**Pull request merged**